src/Indexer.php

```diff
@@ -0,0 +1,222 @@
+<?php
+declare(strict_types = 1);
+
+namespace LanguageServer;
+
+use LanguageServer\Cache\Cache;
+use LanguageServer\FilesFinder\FilesFinder;
+use LanguageServer\Index\{DependenciesIndex, Index};
+use LanguageServer\Protocol\MessageType;
+use Webmozart\PathUtil\Path;
+use Composer\Semver\VersionParser;
+use Sabre\Event\Promise;
+use function Sabre\Event\coroutine;
+
+class Indexer
+{
+    /**
+     * @var The prefix for every cache item
+     */
+    const CACHE_VERSION = 1;
+
+    /**
+     * @var FilesFinder
+     */
+    private $filesFinder;
+
+    /**
+     * @var string
+     */
+    private $rootPath;
+
+    /**
+     * @var LanguageClient
+     */
+    private $client;
+
+    /**
+     * @var Cache
+     */
+    private $cache;
+
+    /**
+     * @var DependenciesIndex
+     */
+    private $dependenciesIndex;
+
+    /**
+     * @var Index
+     */
+    private $sourceIndex;
+
+    /**
+     * @var PhpDocumentLoader
+     */
+    private $documentLoader;
+
+    /**
+     * @var \stdClasss
+     */
+    private $composerLock;
+
+    /**
+     * @param FilesFinder       $filesFinder
+     * @param string            $rootPath
+     * @param LanguageClient    $client
+     * @param Cache             $cache
+     * @param DependenciesIndex $dependenciesIndex
+     * @param Index             $sourceIndex
+     * @param PhpDocumentLoader $documentLoader
+     * @param \stdClass|null    $composerLock
+     */
+    public function __construct(
+        FilesFinder $filesFinder,
+        string $rootPath,
+        LanguageClient $client,
+        Cache $cache,
+        DependenciesIndex $dependenciesIndex,
+        Index $sourceIndex,
+        PhpDocumentLoader $documentLoader,
+        \stdClass $composerLock = null
+    ) {
+        $this->filesFinder = $filesFinder;
+        $this->rootPath = $rootPath;
+        $this->client = $client;
+        $this->cache = $cache;
+        $this->dependenciesIndex = $dependenciesIndex;
+        $this->sourceIndex = $sourceIndex;
+        $this->documentLoader = $documentLoader;
+        $this->composerLock = $composerLock;
+    }
+
+    /**
+     * Will read and parse the passed source files in the project and add them to the appropiate indexes
+     *
+     * @return Promise <void>
+     */
+    public function index(): Promise
+    {
+        return coroutine(function () {
+
+            $pattern = Path::makeAbsolute('**/*.php', $this->rootPath);
+            $uris = yield $this->filesFinder->find($pattern);
+
+            $count = count($uris);
+            $startTime = microtime(true);
+            $this->client->window->logMessage(MessageType::INFO, "$count files total");
+
+            /** @var string[] */
+            $source = [];
+            /** @var string[][] */
+            $deps = [];
+            foreach ($uris as $uri) {
+                if ($this->composerLock !== null && preg_match('/\/vendor\/([^\/]+\/[^\/]+)\//', $uri, $matches)) {
```

Comment:
I think this directory should be pulled from `composer.json`--not everyone uses the default `./vendor/` directory and they are missing out on these wonderful cache benefits.